### PR TITLE
Add logic to export annotations in HTML format

### PR DESCRIPTION
**Pull request by acelaya**

### CODE REVIEW

File: src/sidebar/components/ShareDialog/ExportAnnotations.tsx
Comment (acelaya) [inline src/sidebar/components/ShareDialog/ExportAnnotations.tsx:60]: TBD

File: src/sidebar/services/annotations-exporter.ts
Comment (acelaya) [inline src/sidebar/services/annotations-exporter.ts:92]: I just moved the `join` call some lines below, which made prettier change the indentation to this block

File: src/sidebar/services/test/annotations-exporter-test.js
Comment (acelaya) [inline src/sidebar/services/test/annotations-exporter-test.js:339]: Produced indentation here is not great. However, it's not easy to have proper indentation of all nested elements without using a template engine.

I'm thinking on leaving it like this for now, as indentation does not affect how it renders in the browser, and it's not completely broken when reading it as plain text either.

I'll create a ticket to investigate on this later.

File: src/sidebar/components/ShareDialog/ExportAnnotations.tsx
Comment (acelaya) [inline src/sidebar/components/ShareDialog/ExportAnnotations.tsx:60]: I choose this  description as I don't see how one would import the HTML anywhere. Instead I think this suggests you can open the HTML and see a rich text document that you can then use however you like.

File: src/sidebar/services/annotations-exporter.ts
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.ts:202]: Rather than doing the escaping manually for every element, you can move that into the template string, you can move that into the tag function for the template string.

Something we could alternatively consider using is https://github.com/preactjs/preact-render-to-string, since it runs in the browser. This would mean we write the HTML the same way we author UI components, and render it using this library.

Comment (acelaya) [reply src/sidebar/services/annotations-exporter.ts:202]: > Rather than doing the escaping manually for every element, you can move that into the template string, you can move that into the tag function for the template string.

We can't do that, because it would then escape our own hardcoded tags that we want to preserve. Unless I'm not understanding what you mean.

> Something we could alternatively consider using is https://github.com/preactjs/preact-render-to-string, since it runs in the browser. This would mean we write the HTML the same way we author UI components, and render it using this library.

I like this option, but I was trying to keep the output as human-friendly as possible. If it allows it, I think it's not a bad option at all.

Comment (acelaya) [reply src/sidebar/services/annotations-exporter.ts:202]: > > Rather than doing the escaping manually for every element, you can move that into the template string, you can move that into the tag function for the template string.
> 
> We can't do that, because it would then escape our own hardcoded tags that we want to preserve. Unless I'm not understanding what you mean.

Oh, I was in fact not understanding what you meant. I just got it 😅. Yeah, it makes sense.

File: src/sidebar/services/test/annotations-exporter-test.js
Comment (acelaya) [inline src/sidebar/services/test/annotations-exporter-test.js:367]: My IDE insists in reformatting tabs into spaces every time I save this file.

I thought others could have the same problem, so I "hardcoded" tabs as `\t` to prevent that in the future.

**Comment by acelaya** [comment]
@robertknight I have refactored the whole thing to use `preact-render-to-string`

**Comment by acelaya** [comment]
Hmm, looks like `preact-render-to-string` has a broken type definition file, and we are apparently checking it. I'll investigate this.

**Comment by acelaya** [comment]
> Hmm, looks like `preact-render-to-string` has a broken type definition file, and we are apparently checking it. I'll investigate this.

I can only see one immediate solution to this. Set `"skipLibCheck": true` in tsconfig https://www.typescriptlang.org/tsconfig#skipLibCheck, which is used to skip type-checking third party definitions.

We discussed this some time ago and ended up not doing it because the issue we had with another dependency was solved.

Another mid-term solution would be to provide a PR to the upstream project, trying to fix the type definition, but I would prefer if this PR wasn't blocked by this.

**Comment by robertknight** [comment]
> Another mid-term solution would be to provide a PR to the upstream project, trying to fix the type definition, but I would prefer if this PR wasn't blocked by this.

I'm looking at https://github.com/preactjs/preact-render-to-string/blob/main/jsx.d.ts and wondering how that ever worked. I've pinged the Preact developers in their Slack to ask what the proper fix is.

`skipLibCheck` is unfortunately a global bandaid. I'd prefer if we could find a fix targeted to this specific package.

Based on https://stackoverflow.com/a/41641001/434243 (docs: https://www.typescriptlang.org/tsconfig#paths) I was able to get the code typechecking with this diff:

```diff
diff --git a/src/tsconfig.json b/src/tsconfig.json
index 4f413c4fa..21429e112 100644
--- a/src/tsconfig.json
+++ b/src/tsconfig.json
@@ -21,7 +21,11 @@
     // Prevent automatic inclusion of global variables defined in `@types/<name>` packages.
     // This prevents eg. Node globals from `@types/node` being included when writing
     // code for the browser.
-    "types": []
+    "types": [],
+
+    "paths": {
+      "preact-render-to-string/jsx": ["./types/preact-render-to-string"]
+    }
   },
   "include": ["**/*.js", "**/*.ts", "**/*.tsx", "types/*.d.ts"],
   "exclude": [
diff --git a/src/types/preact-render-to-string.d.ts b/src/types/preact-render-to-string.d.ts
new file mode 100644
index 000000000..09b4fad73
--- /dev/null
+++ b/src/types/preact-render-to-string.d.ts
@@ -0,0 +1,21 @@
+declare module "preact-render-to-string/jsx" {
+  import { VNode } from 'preact';
+
+  interface Options {
+    jsx?: boolean;
+    xml?: boolean;
+    pretty?: boolean | string;
+    shallow?: boolean;
+    functions?: boolean;
+    functionNames?: boolean;
+    skipFalseAttributes?: boolean;
+  }
+
+  export default function renderToStringPretty(
+    vnode: VNode,
+    context?: any,
+    options?: Options
+  ): string;
+
+  export function shallowRender(vnode: VNode, context?: any): string;
+}
```

**Comment by acelaya** [comment]
> Based on https://stackoverflow.com/a/41641001/434243 (docs: https://www.typescriptlang.org/tsconfig#paths) I was able to get the code typechecking with this diff:
> 
> ```diff
> diff --git a/src/tsconfig.json b/src/tsconfig.json
> index 4f413c4fa..21429e112 100644
> --- a/src/tsconfig.json
> +++ b/src/tsconfig.json
> @@ -21,7 +21,11 @@
>      // Prevent automatic inclusion of global variables defined in `@types/<name>` packages.
>      // This prevents eg. Node globals from `@types/node` being included when writing
>      // code for the browser.
> -    "types": []
> +    "types": [],
> +
> +    "paths": {
> +      "preact-render-to-string/jsx": ["./types/preact-render-to-string"]
> +    }
>    },
>    "include": ["**/*.js", "**/*.ts", "**/*.tsx", "types/*.d.ts"],
>    "exclude": [
> diff --git a/src/types/preact-render-to-string.d.ts b/src/types/preact-render-to-string.d.ts
> new file mode 100644
> index 000000000..09b4fad73
> --- /dev/null
> +++ b/src/types/preact-render-to-string.d.ts
> @@ -0,0 +1,21 @@
> +declare module "preact-render-to-string/jsx" {
> +  import { VNode } from 'preact';
> +
> +  interface Options {
> +    jsx?: boolean;
> +    xml?: boolean;
> +    pretty?: boolean | string;
> +    shallow?: boolean;
> +    functions?: boolean;
> +    functionNames?: boolean;
> +    skipFalseAttributes?: boolean;
> +  }
> +
> +  export default function renderToStringPretty(
> +    vnode: VNode,
> +    context?: any,
> +    options?: Options
> +  ): string;
> +
> +  export function shallowRender(vnode: VNode, context?: any): string;
> +}
> ```
> 
> The modified types are the same as the original minus the double default-export.

This works for me as an interim solution. I'll push the changes if you are ok with it.

**Comment by robertknight** [comment]
> This works for me as an interim solution. I'll push the changes if you are ok with it.

Please do.

File: src/sidebar/components/ShareDialog/ExportAnnotations.tsx
Comment (robertknight) [reply src/sidebar/components/ShareDialog/ExportAnnotations.tsx:60]: Now that we have both TXT and HTML,  I think the important distinction to communicate is plain vs rich text. In Slack I suggested something like this:

Text formats

For the purposes of this PR, you could use "For import into word processors as plain text" and "For import into word processors as rich text" as the descriptions of "TXT" and "HTML" respectively which follow the existing structure.

**Comment by acelaya** [comment]
> The HTML output is missing a charset which can lead to rendering issues like this:
> 
> HTML encoding issues
> I suggest the output should have a proper HTML container (`<html>`, `<body>`, doctype, meta tags) so that it renders correctly if you directly open the content in a browser.

Makes sense. Thanks!

File: src/sidebar/services/annotations-exporter.tsx
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.tsx:79]:
```suggestion
    const annotationsAsText = annotations.map((annotation, index) => {
```

File: src/sidebar/services/annotations-exporter.tsx
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.tsx:212]: Per notes on https://developer.mozilla.org/en-US/docs/Web/HTML/Element/article, things that look like headings should really use an `<hN>` tag.

File: src/sidebar/services/annotations-exporter.tsx
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.tsx:179]: The `<header>` element is normally used for a navigation banner on a website. I think this is probably better done as a section, so you'd have something like:

```
<section>
  <h1>Summary</h1>
  ...
</section>
<section>
  <h1>Annotations</h1>
  <article>
    <h2>Annotation 1</h2>
    ...
  </article>
</section>
```

File: src/sidebar/services/annotations-exporter.tsx
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.tsx:228]: What is the `<br />` for? `<article>` elements are blocks so there should be a gap between them automatically.

File: src/sidebar/services/test/annotations-exporter-test.js
Comment (robertknight) [inline src/sidebar/services/test/annotations-exporter-test.js:300]: This is going to be annoying to update if the format ever changes in an unimportant way. It might be more practical to strip formatting from the actual/expected output and just compare that.

File: src/types/preact-render-to-string.d.ts
Comment (robertknight) [inline src/types/preact-render-to-string.d.ts:2]: Would you mind filing an issue upstream with the error and referencing it here. I think it should be easy enough to address.

**Review by robertknight — COMMENTED** [review]
I took a pass over this and left some notes on the HTML structure and a few other details.

File: src/sidebar/services/annotations-exporter.tsx
Comment (acelaya) [reply src/sidebar/services/annotations-exporter.tsx:228]: Just to add some extra space between every annotation block, otherwise there's the same space between every line in one annotation, and the first line on the next one.

On the other hand, probably after applying [this suggestion](https://github.com/hypothesis/client/pull/6073#discussion_r1448721059), this becomes less problematic and we can get rid of the `<br />`s

File: src/sidebar/services/test/annotations-exporter-test.js
Comment (acelaya) [reply src/sidebar/services/test/annotations-exporter-test.js:300]: > It might be more practical to strip formatting from the actual/expected output and just compare that.

Can you elaborate on this?

File: src/types/preact-render-to-string.d.ts
Comment (acelaya) [reply src/types/preact-render-to-string.d.ts:2]: On my way

File: src/sidebar/components/ShareDialog/ExportAnnotations.tsx
Comment (acelaya) [reply src/sidebar/components/ShareDialog/ExportAnnotations.tsx:60]: I'll update the descriptions in this PR, and apply the other improvements once this has been merged. I already created [an item in the board](https://github.com/orgs/hypothesis/projects/131/views/1?pane=issue&itemId=49676664) to make sure this is not lost.

File: src/sidebar/services/test/annotations-exporter-test.js
Comment (robertknight) [reply src/sidebar/services/test/annotations-exporter-test.js:300]: I was thinking of doing `assert.equal(simplifyFormatting(result), simplifyFormatting(expected))` where `simplifyFormatting` is some suitable function (not literally with that name) that normalizes the arguments to remove unimportant whitespace, eg. by stripping leading / trailing spaces from each line.

Comment (acelaya) [reply src/sidebar/services/test/annotations-exporter-test.js:300]: oh, gotcha. Yeah, I think that could work.

File: src/sidebar/services/annotations-exporter.tsx
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.tsx:203]: `<dl>` is semantically appropriate, but seems the import into Word / Google Docs is not ideal. We might be better off switching to a `<table>` for this.

File: src/sidebar/services/annotations-exporter.tsx
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.tsx:222]: I would prefer that we order the content here roughly in the same way that we present it on annotation cards: Metadata (author, timestamp), then quote, then comment, then tags.

File: src/sidebar/services/annotations-exporter.tsx
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.tsx:174]: How about something like `Annotations on ${document title}` as the title. Otherwise every export gets the same tab title.

File: src/sidebar/services/annotations-exporter.tsx
Comment (robertknight) [inline src/sidebar/services/annotations-exporter.tsx:181]: I don't think we should tackle this in this PR, but I noticed when testing this in VoiceOver that it would read a date such as "2023-12-01 14:05" as "two thousand twenty-three to two-thousand twelve", which is wrong.

**Review by robertknight — APPROVED** [review]
I added some notes on the markup structure, as I think there are some elements we will want to change to improve the result of importing it into Word / Google Docs etc. Also I think we could improve the presentation when opened standalone in a browser. Merging this PR will get the basic infrastructure in place though and we can let others start to try it out and provide feedback.

File: src/sidebar/services/annotations-exporter.tsx
Comment (acelaya) [reply src/sidebar/services/annotations-exporter.tsx:222]: Yeah, there's another item to address that, both for HTML and Text.

File: src/sidebar/services/annotations-exporter.tsx
Comment (acelaya) [reply src/sidebar/services/annotations-exporter.tsx:181]: Hmm. I would have expected the use of `time` with the `timestamp` attribute to be used by screen readers to know that's a date, but clearly something is missing.

I'll give it a thought.

**Comment by acelaya** [comment]
> Merging this PR will get the basic infrastructure in place though and we can let others start to try it out and provide feedback.

Yeah, there's a lot of iteration still needed here (not just for HTML, but all formats), but wanted to have something landed soon to gather more feedback.

File: src/sidebar/services/annotations-exporter.tsx
Comment (acelaya) [reply src/sidebar/services/annotations-exporter.tsx:203]: I think a table should work nicely here.

**Pull request merged**
